### ISSUE TITLE
Create utility abstractions to better distribute invariants

### ISSUE THREAD

**Issue by Robbepop** [Robbepop/prophet#6]
Create utility abstractions that help to provide guarantees for invariants throughout the execution of the program. Also testability should be significantly improved by partitioning the program complexity into smaller tiles.

This includes but is not limited to following utility components:

- [x] `WeightsMatrix` & `DeltaWeightsMatrix`: Abstraction built on top of `Array2<f32>` to more semantically provide an interface for weights matrices mainly used by `FullyConnectedLayer`.
- [x] `SignalBuffer`: Abstraction to model signals instead of using raw `Array1<f32>`.
- [x] `ErrorSignalBuffer`: Same as with `SignalBuffer` but specialized for gradients in gradient descent computation.

Certain layer kinds:

- [x] `FullyConnectedLayer`: Abstracts mechanics of fully connected (dense) layers.
- [x] `ActivationLayer`: A layer for activation application.
- [x] `ContainerLayer`: A layer that contains other layers in a strictly sequential order.
- [x] `Layer`, including variants for `FullyConnectedLayer` and `ActivationLayer`

Core functionality & utilities:

- [x] `NeuralNet`: Implements the concrete interface for neural networks and contains layers that describe its computation.

Also model some traits that each stand uniquely for an important operation and behaviour shared by those abstractions:

- [x] `ProcessInputSignal`: For internal (layer-based) feed forward operations.
- [x] `HasOutputSignal`: For layers with output signals. In theory every layer requires this.
- [x] `CalculateOutputErrorSignal`: For layers that work with gradient descent learning.
- [x] `HasErrorSignal`: See above ...
- [x] `PropagateErrorSignal`: See above ...
- [x] `ApplyErrorSignalCorrection`: See above ...
- [x] `SizedLayer`: Interface for working with input and output size of layers.
